File: backend/src/main/kotlin/mymoney/delivery/http/dto/SyncDto.kt
package mymoney.delivery.http.dto

import kotlinx.datetime.Instant
import kotlinx.serialization.Serializable
import mymoney.domain.model.Account
import mymoney.domain.model.Budget
import mymoney.domain.model.BudgetPeriodType
import mymoney.domain.model.Category
import mymoney.domain.model.CategoryType
import mymoney.domain.model.Debt
import mymoney.domain.model.DebtDirection
import mymoney.domain.model.DebtPayment
import mymoney.domain.model.DebtStatus
import mymoney.domain.model.Family
import mymoney.domain.model.FamilyMember
import mymoney.domain.model.FamilyRole
import mymoney.domain.model.Goal
import mymoney.domain.model.Subscription
import mymoney.domain.model.SubscriptionPeriod
import mymoney.domain.model.SyncBundle
import mymoney.domain.model.SyncPullResult
import mymoney.domain.model.SyncPushOutcome
import mymoney.domain.model.Transaction
import mymoney.domain.model.TransactionType
import java.util.UUID

/**
 * DTOs used only by /v1/sync/{pull,push}. Separate from the CRUD DTOs on
 * purpose — sync must transport every column verbatim, while CRUD DTOs stay
 * shaped for the specific request/response they serve.
 */

@Serializable
data class SyncAccountDto(
    val id: String,
    val familyId: String,
    val name: String,
    val type: String,
    val currency: String,
    val initialBalanceKopecks: Long,
    val creditLimitKopecks: Long? = null,
    val isArchived: Boolean,
    val isDeleted: Boolean,
    val createdAt: Instant,
    val updatedAt: Instant,
)

@Serializable
data class SyncCategoryDto(
    val id: String,
    val familyId: String,
    val parentCategoryId: String? = null,
    val name: String,
    val type: String,
    val isMandatory: Boolean,
    val isSystem: Boolean,
    val icon: String? = null,
    val color: String? = null,
    val isArchived: Boolean,
    val isDeleted: Boolean,
    val createdAt: Instant,
    val updatedAt: Instant,
)

@Serializable
data class SyncTransactionDto(
    val id: String,
    val familyId: String,
    val accountId: String,
    val categoryId: String? = null,
    val type: String,
    val targetAccountId: String? = null,
    val amountKopecks: Long,
    val currency: String,
    val occurredAt: Instant,
    val comment: String? = null,
    val attachmentPhotoPath: String? = null,
    val createdBy: String,
    val createdAt: Instant,
    val updatedAt: Instant,
    val isDeleted: Boolean,
)

@Serializable
data class SyncBudgetDto(
    val id: String,
    val familyId: String,
    val categoryId: String,
    val periodType: String,
    val periodStart: Instant,
    val plannedAmountKopecks: Long,
    val isDeleted: Boolean,
    val createdAt: Instant,
    val updatedAt: Instant,
)

@Serializable
data class SyncGoalDto(
    val id: String,
    val familyId: String,
    val name: String,
    val targetAmountKopecks: Long,
    val currentAmountKopecks: Long,
    val targetDate: Instant? = null,
    val isDeleted: Boolean,
    val createdAt: Instant,
    val updatedAt: Instant,
)

@Serializable
data class SyncFamilyDto(
    val id: String,
    val name: String,
    val createdAt: Instant,
    val updatedAt: Instant,
    val isDeleted: Boolean,
)

@Serializable
data class SyncFamilyMemberDto(
    val id: String,
    val familyId: String,
    val userId: String,
    val role: String,
    val joinedAt: Instant,
    val isDeleted: Boolean,
)

@Serializable
data class SyncDebtDto(
    val id: String,
    val familyId: String,
    val counterpartyName: String,
    val direction: String,
    val amountKopecks: Long,
    val interestRate: Double = 0.0,
    val dueDate: Instant? = null,
    val status: String,
    val isDeleted: Boolean,
    val createdAt: Instant,
    val updatedAt: Instant,
)

@Serializable
data class SyncDebtPaymentDto(
    val id: String,
    val debtId: String,
    val dueDate: Instant,
    val plannedAmountKopecks: Long,
    val isPaid: Boolean,
    val paidAt: Instant? = null,
    val transactionId: String? = null,
    val isDeleted: Boolean,
    val createdAt: Instant,
    val updatedAt: Instant,
)

@Serializable
data class SyncSubscriptionDto(
    val id: String,
    val familyId: String,
    val name: String,
    val amountKopecks: Long,
    val billingPeriod: String,
    val nextChargeDate: Instant,
    val categoryId: String? = null,
    val isDeleted: Boolean,
    val createdAt: Instant,
    val updatedAt: Instant,
)

@Serializable
data class SyncBundleDto(
    val accounts: List<SyncAccountDto> = emptyList(),
    val categories: List<SyncCategoryDto> = emptyList(),
    val transactions: List<SyncTransactionDto> = emptyList(),
    val budgets: List<SyncBudgetDto> = emptyList(),
    val goals: List<SyncGoalDto> = emptyList(),
    val debts: List<SyncDebtDto> = emptyList(),
    val debtPayments: List<SyncDebtPaymentDto> = emptyList(),
    val subscriptions: List<SyncSubscriptionDto> = emptyList(),
    val families: List<SyncFamilyDto> = emptyList(),
    val familyMembers: List<SyncFamilyMemberDto> = emptyList(),
)

@Serializable
data class SyncPullResponse(
    val serverTime: Instant,
    val bundle: SyncBundleDto,
)

@Serializable
data class SyncPushRequest(
    val bundle: SyncBundleDto,
)

@Serializable
data class SyncAcceptedRefDto(val table: String, val id: String)

@Serializable
data class SyncConflictDto(
    val table: String,
    val id: String,
    val serverBundle: SyncBundleDto,
)

@Serializable
data class SyncPushResponse(
    val serverTime: Instant,
    val accepted: List<SyncAcceptedRefDto>,
    val conflicts: List<SyncConflictDto>,
)

// --- Mappers: domain → DTO --------------------------------------------------

fun Account.toSyncDto() = SyncAccountDto(
    id = id.toString(),
    familyId = familyId.toString(),
    name = name,
    type = type,
    currency = currency,
    initialBalanceKopecks = initialBalanceKopecks,
    creditLimitKopecks = creditLimitKopecks,
    isArchived = isArchived,
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun Category.toSyncDto() = SyncCategoryDto(
    id = id.toString(),
    familyId = familyId.toString(),
    parentCategoryId = parentCategoryId?.toString(),
    name = name,
    type = type.name,
    isMandatory = isMandatory,
    isSystem = isSystem,
    icon = icon,
    color = color,
    isArchived = isArchived,
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun Transaction.toSyncDto() = SyncTransactionDto(
    id = id.toString(),
    familyId = familyId.toString(),
    accountId = accountId.toString(),
    categoryId = categoryId?.toString(),
    type = type.name,
    targetAccountId = targetAccountId?.toString(),
    amountKopecks = amountKopecks,
    currency = currency,
    occurredAt = occurredAt,
    comment = comment,
    attachmentPhotoPath = attachmentPhotoPath,
    createdBy = createdBy.toString(),
    createdAt = createdAt,
    updatedAt = updatedAt,
    isDeleted = isDeleted,
)

fun Budget.toSyncDto() = SyncBudgetDto(
    id = id.toString(),
    familyId = familyId.toString(),
    categoryId = categoryId.toString(),
    periodType = periodType.name,
    periodStart = periodStart,
    plannedAmountKopecks = plannedAmountKopecks,
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun Debt.toSyncDto() = SyncDebtDto(
    id = id.toString(),
    familyId = familyId.toString(),
    counterpartyName = counterpartyName,
    direction = direction.name,
    amountKopecks = amountKopecks,
    interestRate = interestRate,
    dueDate = dueDate,
    status = status.name,
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun DebtPayment.toSyncDto() = SyncDebtPaymentDto(
    id = id.toString(),
    debtId = debtId.toString(),
    dueDate = dueDate,
    plannedAmountKopecks = plannedAmountKopecks,
    isPaid = isPaid,
    paidAt = paidAt,
    transactionId = transactionId?.toString(),
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun Subscription.toSyncDto() = SyncSubscriptionDto(
    id = id.toString(),
    familyId = familyId.toString(),
    name = name,
    amountKopecks = amountKopecks,
    billingPeriod = billingPeriod.name,
    nextChargeDate = nextChargeDate,
    categoryId = categoryId?.toString(),
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun Goal.toSyncDto() = SyncGoalDto(
    id = id.toString(),
    familyId = familyId.toString(),
    name = name,
    targetAmountKopecks = targetAmountKopecks,
    currentAmountKopecks = currentAmountKopecks,
    targetDate = targetDate,
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun Family.toSyncDto() = SyncFamilyDto(
    id = id.toString(),
    name = name,
    createdAt = createdAt,
    updatedAt = updatedAt,
    isDeleted = isDeleted,
)

fun FamilyMember.toSyncDto() = SyncFamilyMemberDto(
    id = id.toString(),
    familyId = familyId.toString(),
    userId = userId.toString(),
    role = role.name,
    joinedAt = joinedAt,
    isDeleted = isDeleted,
)

fun SyncBundle.toDto() = SyncBundleDto(
    accounts = accounts.map { it.toSyncDto() },
    categories = categories.map { it.toSyncDto() },
    transactions = transactions.map { it.toSyncDto() },
    budgets = budgets.map { it.toSyncDto() },
    goals = goals.map { it.toSyncDto() },
    debts = debts.map { it.toSyncDto() },
    debtPayments = debtPayments.map { it.toSyncDto() },
    subscriptions = subscriptions.map { it.toSyncDto() },
    families = families.map { it.toSyncDto() },
    familyMembers = familyMembers.map { it.toSyncDto() },
)

fun SyncPullResult.toResponse() = SyncPullResponse(
    serverTime = serverTime,
    bundle = bundle.toDto(),
)

fun SyncPushOutcome.toResponse() = SyncPushResponse(
    serverTime = serverTime,
    accepted = accepted.map { SyncAcceptedRefDto(it.table, it.id) },
    conflicts = conflicts.map {
        SyncConflictDto(table = it.table, id = it.id, serverBundle = it.serverBundle.toDto())
    },
)

// --- Mappers: DTO → domain --------------------------------------------------

fun SyncAccountDto.toDomain() = Account(
    id = UUID.fromString(id),
    familyId = UUID.fromString(familyId),
    name = name,
    type = type,
    currency = currency,
    initialBalanceKopecks = initialBalanceKopecks,
    creditLimitKopecks = creditLimitKopecks,
    isArchived = isArchived,
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun SyncCategoryDto.toDomain() = Category(
    id = UUID.fromString(id),
    familyId = UUID.fromString(familyId),
    parentCategoryId = parentCategoryId?.let { UUID.fromString(it) },
    name = name,
    type = CategoryType.valueOf(type),
    isMandatory = isMandatory,
    isSystem = isSystem,
    icon = icon,
    color = color,
    isArchived = isArchived,
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun SyncTransactionDto.toDomain() = Transaction(
    id = UUID.fromString(id),
    familyId = UUID.fromString(familyId),
    accountId = UUID.fromString(accountId),
    categoryId = categoryId?.let { UUID.fromString(it) },
    type = TransactionType.valueOf(type),
    targetAccountId = targetAccountId?.let { UUID.fromString(it) },
    amountKopecks = amountKopecks,
    currency = currency,
    occurredAt = occurredAt,
    comment = comment,
    attachmentPhotoPath = attachmentPhotoPath,
    createdBy = UUID.fromString(createdBy),
    createdAt = createdAt,
    updatedAt = updatedAt,
    isDeleted = isDeleted,
)

fun SyncBudgetDto.toDomain() = Budget(
    id = UUID.fromString(id),
    familyId = UUID.fromString(familyId),
    categoryId = UUID.fromString(categoryId),
    periodType = BudgetPeriodType.valueOf(periodType),
    periodStart = periodStart,
    plannedAmountKopecks = plannedAmountKopecks,
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun SyncGoalDto.toDomain() = Goal(
    id = UUID.fromString(id),
    familyId = UUID.fromString(familyId),
    name = name,
    targetAmountKopecks = targetAmountKopecks,
    currentAmountKopecks = currentAmountKopecks,
    targetDate = targetDate,
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun SyncFamilyDto.toDomain() = Family(
    id = UUID.fromString(id),
    name = name,
    createdAt = createdAt,
    updatedAt = updatedAt,
    isDeleted = isDeleted,
)

fun SyncFamilyMemberDto.toDomain() = FamilyMember(
    id = UUID.fromString(id),
    familyId = UUID.fromString(familyId),
    userId = UUID.fromString(userId),
    role = FamilyRole.valueOf(role),
    joinedAt = joinedAt,
    isDeleted = isDeleted,
)

fun SyncDebtDto.toDomain() = Debt(
    id = UUID.fromString(id),
    familyId = UUID.fromString(familyId),
    counterpartyName = counterpartyName,
    direction = DebtDirection.valueOf(direction),
    amountKopecks = amountKopecks,
    interestRate = interestRate,
    dueDate = dueDate,
    status = DebtStatus.valueOf(status),
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun SyncDebtPaymentDto.toDomain() = DebtPayment(
    id = UUID.fromString(id),
    debtId = UUID.fromString(debtId),
    dueDate = dueDate,
    plannedAmountKopecks = plannedAmountKopecks,
    isPaid = isPaid,
    paidAt = paidAt,
    transactionId = transactionId?.let { UUID.fromString(it) },
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun SyncSubscriptionDto.toDomain() = Subscription(
    id = UUID.fromString(id),
    familyId = UUID.fromString(familyId),
    name = name,
    amountKopecks = amountKopecks,
    billingPeriod = SubscriptionPeriod.valueOf(billingPeriod),
    nextChargeDate = nextChargeDate,
    categoryId = categoryId?.let { UUID.fromString(it) },
    isDeleted = isDeleted,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun SyncBundleDto.toDomain() = SyncBundle(
    accounts = accounts.map { it.toDomain() },
    categories = categories.map { it.toDomain() },
    transactions = transactions.map { it.toDomain() },
    budgets = budgets.map { it.toDomain() },
    goals = goals.map { it.toDomain() },
    debts = debts.map { it.toDomain() },
    debtPayments = debtPayments.map { it.toDomain() },
    subscriptions = subscriptions.map { it.toDomain() },
    families = families.map { it.toDomain() },
    familyMembers = familyMembers.map { it.toDomain() },
)

File: mobile/lib/data/remote/dto/sync_dto.dart
import '../../../domain/model/account.dart';
import '../../../domain/model/budget.dart';
import '../../../domain/model/category.dart';
import '../../../domain/model/enums.dart';
import '../../../domain/model/goal.dart';
import '../../../domain/model/transaction.dart';

/// Wire-level bundle exchanged with /v1/sync/pull and /v1/sync/push.
///
/// Field order and JSON shape must stay lock-step with the Kotlin
/// SyncBundleDto (see backend/.../delivery/http/dto/SyncDto.kt). If either
/// side drifts, sync silently corrupts data.
class SyncBundleDto {
  const SyncBundleDto({
    this.accounts = const [],
    this.categories = const [],
    this.transactions = const [],
    this.budgets = const [],
    this.goals = const [],
  });

  final List<Account> accounts;
  final List<Category> categories;
  final List<Transaction> transactions;
  final List<Budget> budgets;
  final List<Goal> goals;

  bool get isEmpty =>
      accounts.isEmpty &&
      categories.isEmpty &&
      transactions.isEmpty &&
      budgets.isEmpty &&
      goals.isEmpty;

  Map<String, dynamic> toJson() => {
        'accounts': accounts.map(accountToJson).toList(),
        'categories': categories.map(categoryToJson).toList(),
        'transactions': transactions.map(transactionToJson).toList(),
        'budgets': budgets.map(budgetToJson).toList(),
        'goals': goals.map(goalToJson).toList(),
        'families': const <Map<String, dynamic>>[],
        'familyMembers': const <Map<String, dynamic>>[],
      };

  static SyncBundleDto fromJson(Map<String, dynamic> json) => SyncBundleDto(
        accounts: (json['accounts'] as List<dynamic>? ?? const [])
            .map((e) => accountFromJson(e as Map<String, dynamic>))
            .toList(),
        categories: (json['categories'] as List<dynamic>? ?? const [])
            .map((e) => categoryFromJson(e as Map<String, dynamic>))
            .toList(),
        transactions: (json['transactions'] as List<dynamic>? ?? const [])
            .map((e) => transactionFromJson(e as Map<String, dynamic>))
            .toList(),
        budgets: (json['budgets'] as List<dynamic>? ?? const [])
            .map((e) => budgetFromJson(e as Map<String, dynamic>))
            .toList(),
        goals: (json['goals'] as List<dynamic>? ?? const [])
            .map((e) => goalFromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class SyncPullResponse {
  const SyncPullResponse({required this.serverTime, required this.bundle});
  final DateTime serverTime;
  final SyncBundleDto bundle;

  static SyncPullResponse fromJson(Map<String, dynamic> json) => SyncPullResponse(
        serverTime: DateTime.parse(json['serverTime'] as String).toUtc(),
        bundle: SyncBundleDto.fromJson(json['bundle'] as Map<String, dynamic>),
      );
}

class SyncPushResponse {
  const SyncPushResponse({
    required this.serverTime,
    required this.accepted,
    required this.conflicts,
  });
  final DateTime serverTime;
  final List<AcceptedRef> accepted;
  final List<SyncConflict> conflicts;

  static SyncPushResponse fromJson(Map<String, dynamic> json) => SyncPushResponse(
        serverTime: DateTime.parse(json['serverTime'] as String).toUtc(),
        accepted: (json['accepted'] as List<dynamic>? ?? const [])
            .map((e) => AcceptedRef.fromJson(e as Map<String, dynamic>))
            .toList(),
        conflicts: (json['conflicts'] as List<dynamic>? ?? const [])
            .map((e) => SyncConflict.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class AcceptedRef {
  const AcceptedRef({required this.table, required this.id});
  final String table;
  final String id;
  static AcceptedRef fromJson(Map<String, dynamic> json) =>
      AcceptedRef(table: json['table'] as String, id: json['id'] as String);
}

class SyncConflict {
  const SyncConflict({
    required this.table,
    required this.id,
    required this.serverBundle,
  });
  final String table;
  final String id;
  final SyncBundleDto serverBundle;
  static SyncConflict fromJson(Map<String, dynamic> json) => SyncConflict(
        table: json['table'] as String,
        id: json['id'] as String,
        serverBundle: SyncBundleDto.fromJson(json['serverBundle'] as Map<String, dynamic>),
      );
}

// --- Per-entity conversions -------------------------------------------------

Map<String, dynamic> accountToJson(Account a) => {
      'id': a.id,
      'familyId': a.familyId,
      'name': a.name,
      'type': a.type,
      'currency': a.currency,
      'initialBalanceKopecks': a.initialBalanceKopecks,
      'isArchived': a.isArchived,
      'isDeleted': a.isDeleted,
      'createdAt': _iso(a.createdAt),
      'updatedAt': _iso(a.updatedAt),
    };

Account accountFromJson(Map<String, dynamic> j) => Account(
      id: j['id'] as String,
      familyId: j['familyId'] as String,
      name: j['name'] as String,
      type: j['type'] as String,
      currency: j['currency'] as String? ?? 'RUB',
      initialBalanceKopecks: (j['initialBalanceKopecks'] as num).toInt(),
      isArchived: j['isArchived'] as bool? ?? false,
      isDeleted: j['isDeleted'] as bool? ?? false,
      createdAt: DateTime.parse(j['createdAt'] as String).toUtc(),
      updatedAt: DateTime.parse(j['updatedAt'] as String).toUtc(),
    );

Map<String, dynamic> categoryToJson(Category c) => {
      'id': c.id,
      'familyId': c.familyId,
      'parentCategoryId': c.parentCategoryId,
      'name': c.name,
      'type': c.type.code,
      'isMandatory': c.isMandatory,
      'isSystem': c.isSystem,
      'icon': c.icon,
      'color': c.color,
      'isArchived': c.isArchived,
      'isDeleted': c.isDeleted,
      'createdAt': _iso(c.createdAt),
      'updatedAt': _iso(c.updatedAt),
    };

Category categoryFromJson(Map<String, dynamic> j) => Category(
      id: j['id'] as String,
      familyId: j['familyId'] as String,
      parentCategoryId: j['parentCategoryId'] as String?,
      name: j['name'] as String,
      type: _parseCategoryType(j['type'] as String),
      isMandatory: j['isMandatory'] as bool? ?? false,
      isSystem: j['isSystem'] as bool? ?? false,
      icon: j['icon'] as String?,
      color: j['color'] as String?,
      isArchived: j['isArchived'] as bool? ?? false,
      isDeleted: j['isDeleted'] as bool? ?? false,
      createdAt: DateTime.parse(j['createdAt'] as String).toUtc(),
      updatedAt: DateTime.parse(j['updatedAt'] as String).toUtc(),
    );

Map<String, dynamic> transactionToJson(Transaction t) => {
      'id': t.id,
      'familyId': t.familyId,
      'accountId': t.accountId,
      'categoryId': t.categoryId,
      'type': t.type.code,
      'targetAccountId': t.targetAccountId,
      'amountKopecks': t.amountKopecks,
      'currency': t.currency,
      'occurredAt': _iso(t.occurredAt),
      'comment': t.comment,
      'attachmentPhotoPath': t.attachmentPhotoPath,
      'createdBy': t.createdBy,
      'createdAt': _iso(t.createdAt),
      'updatedAt': _iso(t.updatedAt),
      'isDeleted': t.isDeleted,
    };

Transaction transactionFromJson(Map<String, dynamic> j) => Transaction(
      id: j['id'] as String,
      familyId: j['familyId'] as String,
      accountId: j['accountId'] as String,
      categoryId: j['categoryId'] as String?,
      type: _parseTransactionType(j['type'] as String),
      targetAccountId: j['targetAccountId'] as String?,
      amountKopecks: (j['amountKopecks'] as num).toInt(),
      currency: j['currency'] as String? ?? 'RUB',
      occurredAt: DateTime.parse(j['occurredAt'] as String).toUtc(),
      comment: j['comment'] as String?,
      attachmentPhotoPath: j['attachmentPhotoPath'] as String?,
      createdBy: j['createdBy'] as String,
      createdAt: DateTime.parse(j['createdAt'] as String).toUtc(),
      updatedAt: DateTime.parse(j['updatedAt'] as String).toUtc(),
      isDeleted: j['isDeleted'] as bool? ?? false,
    );

Map<String, dynamic> budgetToJson(Budget b) => {
      'id': b.id,
      'familyId': b.familyId,
      'categoryId': b.categoryId,
      'periodType': b.periodType.code,
      'periodStart': _iso(b.periodStart),
      'plannedAmountKopecks': b.plannedAmountKopecks,
      'isDeleted': b.isDeleted,
      'createdAt': _iso(b.createdAt),
      'updatedAt': _iso(b.updatedAt),
    };

Budget budgetFromJson(Map<String, dynamic> j) => Budget(
      id: j['id'] as String,
      familyId: j['familyId'] as String,
      categoryId: j['categoryId'] as String,
      periodType: _parseBudgetPeriodType(j['periodType'] as String),
      periodStart: DateTime.parse(j['periodStart'] as String).toUtc(),
      plannedAmountKopecks: (j['plannedAmountKopecks'] as num).toInt(),
      createdAt: DateTime.parse(j['createdAt'] as String).toUtc(),
      updatedAt: DateTime.parse(j['updatedAt'] as String).toUtc(),
      isDeleted: j['isDeleted'] as bool? ?? false,
    );

Map<String, dynamic> goalToJson(Goal g) => {
      'id': g.id,
      'familyId': g.familyId,
      'name': g.name,
      'targetAmountKopecks': g.targetAmountKopecks,
      'currentAmountKopecks': g.currentAmountKopecks,
      'targetDate': g.targetDate == null ? null : _iso(g.targetDate!),
      'isDeleted': g.isDeleted,
      'createdAt': _iso(g.createdAt),
      'updatedAt': _iso(g.updatedAt),
    };

Goal goalFromJson(Map<String, dynamic> j) => Goal(
      id: j['id'] as String,
      familyId: j['familyId'] as String,
      name: j['name'] as String,
      targetAmountKopecks: (j['targetAmountKopecks'] as num).toInt(),
      currentAmountKopecks: (j['currentAmountKopecks'] as num).toInt(),
      targetDate: j['targetDate'] == null
          ? null
          : DateTime.parse(j['targetDate'] as String).toUtc(),
      createdAt: DateTime.parse(j['createdAt'] as String).toUtc(),
      updatedAt: DateTime.parse(j['updatedAt'] as String).toUtc(),
      isDeleted: j['isDeleted'] as bool? ?? false,
    );

String _iso(DateTime dt) => dt.toUtc().toIso8601String();

CategoryType _parseCategoryType(String s) => switch (s) {
      'INCOME' => CategoryType.income,
      'EXPENSE' => CategoryType.expense,
      _ => throw ArgumentError('Unknown CategoryType: $s'),
    };

TransactionType _parseTransactionType(String s) => switch (s) {
      'INCOME' => TransactionType.income,
      'EXPENSE' => TransactionType.expense,
      'TRANSFER' => TransactionType.transfer,
      _ => throw ArgumentError('Unknown TransactionType: $s'),
    };

BudgetPeriodType _parseBudgetPeriodType(String s) => switch (s) {
      'WEEK' => BudgetPeriodType.week,
      'MONTH' => BudgetPeriodType.month,
      'YEAR' => BudgetPeriodType.year,
      _ => throw ArgumentError('Unknown BudgetPeriodType: $s'),
    };
